### ISSUE TITLE
How to deal with different versions in different environment?

### ISSUE THREAD

**Issue by tonsV2** [roboll/helmfile#460]
I found it difficult dealing with different versions across different environments

I've came up with the following but I'd love to see someone come up with something simpler.

helmfile.yaml:

    environments:
      dev:
        values:
          - env/dev.yaml
      prod:
        values:
          - env/prod.yaml
    
    releases:
      - name: some-name
        namespace: some-namespace
        chart: some/thing
        version: {{ index .Environment.Values.version "some-name" }}

env/dev.yaml:

    version:
      some-name: "0.1"

env/prod.yaml:

    version:
      some-name: "1.0"

Please let me know what you think.

**Comment by manics** [comment]
I used something similar so that I could override values for dev vs production environments:
```
    values:
      - values/jupyterhub-config.yaml
    {{ range .Environment.Values.jupyterhub_public_overrides }}
      - {{ . }}
    {{ end }}
```
https://github.com/IDR/k8s-analysis-deploy/blob/2019.01.30/helmfile.yaml#L104
`jupyterhub_public_overrides` is `[]` for production.

What would be nice is if `index .Environment.Values` could take a default, so helmfile.yml could contain the full production config, with only the dev overrides needing to be defined, e.g.:
- `{{ index .Environment.Values.version "some-name" | default "1.0" }}`
- `{{ range .Environment.Values.jupyterhub_public_overrides | default nil }}`

At the moment this fails if the values aren't defined:
```
error during helmfile.yaml parsing: template: stringTemplate:34:25: executing "stringTemplate" at <.Environment.Values....>: map has no entry for key "jupyterhub_staging_overrides"
```

**Comment by tonsV2** [comment]
So jupyterhub_public_overrides is defined in an environment specific values file as a list of files? But, in your case, empty for production? Could you elaborate a bit more, sorry for being a bit slow.

**Comment by manics** [comment]
I've got two environments with separate values files in that git repo: https://github.com/IDR/k8s-analysis-deploy/blob/2019.01.30/helmfile.yaml#L237
- [`production.yaml`](https://github.com/IDR/k8s-analysis-deploy/blob/2019.01.30/production.yaml) defined the variables as empty
- [`minikube.yaml`](https://github.com/IDR/k8s-analysis-deploy/blob/2019.01.30/minikube.yaml) contains overrides

**Comment by ingvarch** [comment]
Any updates? I have the same issue.

**Comment by sgandon** [comment]
@ingvarch what issue exactly ? this was just a question here.
@tonsV2 I guess you could also use the environment name in the values name like this
```
environments:
  dev:
  prod:

releases:
  - name: some-name
    namespace: some-namespace
    chart: some/thing
    version: {{ index .Environment.Values.version "some-name" }}
    values:
      - {{ .Environment.Name }}/prod.yaml
```
But this is not really much simpler. 
How would you imagine something more simpler though ?

**Comment by sgandon** [comment]
I guess this would be more simpler with this request developped : https://github.com/roboll/helmfile/issues/359

**Comment by mumoshu** [comment]
@manics 

> What would be nice is if index .Environment.Values could take a default, so helmfile.yml could contain the full production config, with only the dev overrides needing to be defined

Would #465 resolves your issue? With #465, I believe you can write it like:

```
# Instead of: {{ index .Environment.Values.version "some-name" | default "1.0" }}

{{ .Environment.Values | get "version.some-name" "1.0" }}
```

**Comment by mumoshu** [comment]
@tonsV2 @ingvarch Hey!

Can I boil this issue down to "We need a less verbose way to achieve defaulting for environment values"?

In that case, 

- As @sgandon has suggested #359 would make defaulting possible without defining extra environment values.yaml files
- #465 would make defaulting possible without defining extra environment values

Do you think we can resolve this after either or both of them are addressed?

**Comment by mumoshu** [comment]
FYI #359 has been finally implemented via #622

**Comment by mumoshu** [comment]
Closing this as resolved because all the suggested solutions are implemented. Please feel free to reopen if necessary!
